crates/wasm-smith/src/lib.rs

```diff
@@ -1373,6 +1390,111 @@ where
         Ok(())
     }
 
+    fn arbitrary_available_imports(
+        &mut self,
+        available_imports: Vec<Import>,
+        min: usize,
+        u: &mut Unstructured,
+    ) -> Result<()> {
+        let mut available_funcs = Vec::new();
+        for import in available_imports {
+            match import.desc {
+                ImportDesc::Func(_) => &mut available_funcs,
+                ImportDesc::Global(_) => todo!(),
+                ImportDesc::Table(_) => todo!(),
+                ImportDesc::Memory(_) => todo!(),
+            }
+            .push((import, Cell::new(None)))
+        }
+
+        let mut choices: Vec<
+            for<'i> fn(
+                &mut Unstructured,
+                &'i [(Import, Cell<Option<u32>>)],
+            ) -> Result<&'i (Import, Cell<Option<u32>>)>,
+        > = Vec::with_capacity(1);
+
+        let mut imports = Vec::new();
+        arbitrary_loop(u, min, self.config.max_imports() - self.num_imports, |u| {
+            choices.clear();
+            if self.can_add_local_or_import_func() && !available_funcs.is_empty() {
+                choices.push(|u, f| u.choose(f))
+            }
+
+            if choices.is_empty() {
+                return Ok(false);
+            }
+
+            let f = u.choose(&choices)?;
+            let (import, idx) = f(u, &available_funcs)?;
+
+            let module = import.module.clone();
+            let name = import.name.clone();
+            let ty: EntityType = match &import.desc {
+                ImportDesc::Func(func) => {
+                    if idx.get().is_none() && self.types.len() < self.config.max_types() {
+                        // Try defining the type in the first section.
+                        // TODO: pick section arbitrary
+                        for (section_idx, section) in self.initial_sections.iter_mut().enumerate() {
+                            if let InitialSection::Type(types) = section {
+                                idx.set(Some(self.types.len() as u32));
+                                self.func_types.push(self.types.len() as u32);
+                                self.types.push(LocalType::Defined {
+                                    section: section_idx,
+                                    nth: types.len(),
+                                });
+                                types.push(Type::Func(Rc::new(func.clone())));
+                                break;
+                            }
+                        }
+                    }
+                    match idx.get() {
+                        Some(idx) => {
+                            let ty = self.func_type(idx).clone();
+                            EntityType::Func(idx, ty)
+                        }
+                        None => return Ok(false),
```

Comment:
In this case we can create a type section and push it onto `self.initial_sections`.